hnn_core/cell.py

```diff
@@ -361,14 +419,48 @@ def _set_biophysics(self, sections):
             for mech_name, p_mech in section.mechs.items():
                 sec.insert(mech_name)
                 for attr, val in p_mech.items():
-                    if hasattr(val, '__call__'):
-                        sec.push()
-                        for seg in sec:
-                            setattr(seg, attr, val(h.distance(seg.x)))
-                        h.pop_section()
+                    if isinstance(val, list):
+                        seg_xs, seg_vals = val[0], val[1]
+                        for seg, seg_x, seg_val in zip(sec, seg_xs, seg_vals):
+                            # Checking equaliy till 5 decimal places
+                            np.testing.assert_almost_equal(seg.x, seg_x, 5)
+                            setattr(seg, attr, seg_val)
                     else:
                         setattr(sec, attr, val)
 
+    def _set_section_mechs(self):
+        sections = self.sections
+        for sec_name, section in sections.items():
+            for mech_name, p_mech in section.mechs.items():
+                for attr, val in p_mech.items():
+                    if hasattr(val, '__call__'):
+                        seg_xs, seg_vals = list(), list()
+                        section_distance = self.distance_section(sec_name,
+                                                                 ('soma', 0))
+                        adjacent_seg_dist = 1 / section.nseg
+                        first_seg_centre = (0.5 - (((section.nseg - 1) / 2) *
+                                            adjacent_seg_dist))
+                        seg_centres = list()
+
+                        # Finding centres of all segments in the section
+                        # If number of segments is 5 then seg_centres will
+                        # be 0.1, 0.3, 0.5, 0.7 and 0.9.
+                        for i in range(0, section.nseg):
+                            seg_centres.append(first_seg_centre +
+                                               (i * adjacent_seg_dist))
```

Comment:
any chance we can use `np.linspace` ? That seems exactly for this purpose

Reply:
Very slight precision errors are there. But it is passing the check in `_set_biophysics()`